galsim/phase_psf.py

```diff
@@ -0,0 +1,1071 @@
+diffraction equation:
+
+PSF(x, y) = int( |FT(aperture(u, v) * exp(i * phase(u, v, x, y, t)))|^2, dt)
+
+where x, y are focal plane coordinates and u, v are pupil plane coordinates
+
+The main classes of note are:
+
+AtmosphericScreen
+  Class implementing phase(u, v, x, y, t) for von Karman type turbulence, with possibly evolving
+  "non-frozen-flow" phases.
+
+PhaseScreenList
+  Python sequence type to hold multiple phase screens, for instance to simulate turbulence at
+  different altitudes.  A key method is makePSF(), which will take the list of phase screens,
+  add them together linearly (Fraunhofer approximation), and evaluate the above diffraction
+  equation to yield a PhaseScreenPSF object.
+
+PhaseScreenPSF
+  A GSObject holding the evaluated PSF.
+
+Atmosphere
+  Convenience function to quickly assemble multiple AtmosphericScreens into a PhaseScreenList.
+"""
+
+import numpy as np
+import galsim
+import utilities
+from galsim import GSObject
+
+
+class Aperture(object):
+    """ Class representing a telescope aperture as part of a larger pupil plane.
+
+        The constructed object has two key attributes:
+            `illuminated`  a boolean array indicating which positions in the pupil plane are exposed
+                           to the sky.
+            `rho`          array of unit disc-scaled pupil coordinates for use by Zernike
+                           polynomials (as a complex number).
+
+        `rho` is
+
+        @param pupil_plane_size  Size of the pupil plane in meters.  Note, this may be (in fact, it
+                                 usually *should* be) larger than the aperture diameter.
+        @param npix              Number of pupil plane resolution elements.
+        @param diam              Aperture diameter in meters. [default: pupil_plane_size]
+        @param circular_pupil    Adopt a circular pupil? [default: True].
+        @param obscuration       Fractional linear circular obscuration of pupil. [default: 0.]
+        @param nstruts           Number of radial support struts to add to the central obscuration.
+                                 [default: 0]
+        @param strut_thick       Thickness of support struts as a fraction of pupil diameter.
+                                 [default: 0.05]
+        @param strut_angle       Angle made between the vertical and the first strut in the CCW
+                                 direction; must be an Angle instance.
+                                 [default: 0. * galsim.degrees]
+    """
+    def __init__(self, pupil_plane_size, npix, diam=None, circular_pupil=True, obscuration=0.,
+                 nstruts=0, strut_thick=0.05, strut_angle=0.*galsim.degrees):
+        if obscuration >= 1.:
+            raise ValueError("Pupil fully obscured! obscuration = {1} (>= 1)".format(obscuration))
+        if diam is None:
+            diam = pupil_plane_size
+        self.pupil_plane_size = float(pupil_plane_size)
+        self.npix = int(npix)
```

Comment:
I'm curious - do you think this is a routine that non-expert users might use?  If so, I think it would be warranted to give some guidance in the docstring on sensible levels of npix to ensure adequate sampling of the pupil plane.  Or, at least check the input value and raise a warning if it seems inadequately sampled.